File: week1/2_Valid_Anagram.py
# Valid Anagram
# Two Pointers
# Time Complexity: O(n)
# Space Complexity: O(1)

def isValidAnagram(s: str, t: str) -> bool:
    if len(s) != len(t):
        return False

    count = {}
    for char in s:
        count[char] = count.get(char, 0) + 1

    for char in t:
        if char not in count:
            return False
        count[char] -= 1
        if count[char] < 0:
            return False

    return True

print(isValidAnagram("anagram", "nagaram"))

File: week1/2_Valid_Anagram.ts
// Valid Anagram
// Hash Table
// Time Complexity: O(n)
// Space Complexity: O(1)

const isValidAnagram = (s: string, t: string): boolean => {
  if (s.length !== t.length) return false;

  const count = new Map();
  for (const char of s) {
    count.set(char, (count.get(char) || 0) + 1);
  }

  for (const char of t) {
    if (!count.has(char)) return false;
    count.set(char, count.get(char) - 1);
    if (count.get(char) < 0) return false;
  }

  return true;
};

console.log(isValidAnagram('anagram', 'nagaram'));
